### ISSUE TITLE
`dotnet new --uninstall` not working for folder templates

### ISSUE THREAD

**Issue by patridge** [dotnet/templating#1226]
[Migrating from dotnet/core#893 per [tweets with @sayedihashimi](https://twitter.com/sayedihashimi/status/901533518957694976)]

# Unable to uninstall folder-based templates

This could be an issue with the `dotnet net --uninstall` (or `dotnet new -u`) command, or just an issue with [uninstall documentation](https://docs.microsoft.com/en-us/dotnet/core/tools/custom-templates#to-uninstall-a-template-from-a-file-system-directory), but it doesn't seem possible to uninstall a template via a folder template.

See also this [Stack Overflow question](https://stackoverflow.com/q/45865134/48700).

# General

## Steps

1. Create template folder (**.template.config** folder with a **template.json** file)

    I have confirmed it with this JSON config.

    ```json
    {
        "$schema": "http://json.schemastore.org/template",
        "author": "Some Author",
        "classifications": [ "Some", "Various", "Search Keywords" ],
        "identity": "Some.Identity",
        "shortName": "name-used-in-dotnetnew",
        "name": "Some Template Name"
    }
    ```

2. [Install template](https://docs.microsoft.com/en-us/dotnet/core/tools/custom-templates#to-install-a-template-from-a-file-system-directory)

    ```bash
    dotnet new --install ./some-template-folder/
    ```

3. [optional] Use newly-installed template

    ```bash
    The template "Some Template Name" was created successfully.
    ```

4. Attempt to [uninstall template](https://docs.microsoft.com/en-us/dotnet/core/tools/custom-templates#to-uninstall-a-template-from-a-file-system-directory)

    ```bash
    dotnet new --uninstall ./some-template-folder/
    ```

    Instead of uninstalling, I keep getting this error message.

    > `Could not find something to uninstall called '.\some-template-folder\'.`

This happened with .NET Core SDK v1.1.0 and v2.0.0.

**Comment by seancpeters** [comment]
@patridge - If you provide the full path to the folder you installed from, the uninstall should work.

**Comment by patridge** [comment]
I tried both a relative path and a fully-qualified path on the `--uinstall` command. Unfortunately, nothing seems to uninstall a template installed from a folder (which can be installed from a relative path).

Failing commands:
* [from above] `dotnet new --uninstall ./some-template-folder/`
    * > Could not find something to uninstall called '.\some-template-folder\'.
* `dotnet new --uninstall "C:\full\path\to\some-template-folder\"`
    * > Could not find something to uninstall called 'C:\full\path\to\some-template-folder\'.
* `dotnet new --uninstall "C:\full\path\to\some-template-folder\.template.config"
    * > Could not find something to uninstall called 'C:\full\path\to\some-template-folder\.template.config'.

**Comment by patridge** [comment]
If it helps to reproduce the issue elsewhere, here is the sample template I'm using to reproduce it on my machine: [some-template-folder.zip](https://github.com/dotnet/templating/files/1261360/some-template-folder.zip)

**Comment by seancpeters** [comment]
@patridge - I believe you have discovered a bug. Could you try uninstalling one more time, using the full path, but without a trailing slash:
`dotnet new --uninstall "C:\full\path\to\some-template-folder"`

Could you also let me know the output from running `dotnet --version`?

**Comment by patridge** [comment]
@seancpeters No luck with that variant either (added to above list of commands).

`dotnet --version` -> `2.0.0`

(I had the same uninstall issue with v1.0.0, but I think I remember reading somewhere that uninstall wasn't actually supported for that version.)

**Comment by seancpeters** [comment]
So far, I've been unable to reproduce this issue with dotnet 2.0.0. 

One possibility is that your dotnet new configuration, may be in an unexpected state due to previous installs of this template. If you want to reset the template installation to the defaults dotnet new ships with, you can run `dotnet new --debug:reinit` (this flag is not technically supported, may change / be removed in the future). After running that command, an empty template list will be displayed, but then the next time you run `dotnet new`, it will go through the first-run code, and setup to be in the initial state. This might be good, to get into a known state before retrying the install / uninstall. But be aware that any other templates you've installed will be removed and need to be re-installed.

Another possibility is that we're subtlely miscommunicating about the paths being used. For my tests, I have a directory structure that looks like:
```
C:\TEMP\TEMP_TEMPLATES
\---some-template-folder
    |   some-file.txt
    |
    \---.template.config
            template.json
```
(output from `c:\temp\temp_templates>tree . /A /F`)

From the `c:\temp\temp_templates\` directory, I run the command:
`dotnet new --install some-template-folder\`
... and observe the newly installed template in the template list.

And then to uninistall, I run:
`dotnet new --uninstall c:\temp\temp_templates\some-template-folder`
... and observe the template has been removed from the list.

**Comment by patridge** [comment]
I should be able to test this out tonight or later this weekend.

@seancpeters I couldn't seem to find where the installed templates are actually stored due to `dotnet new` typical-use noise in search results. Is there somewhere in the file system, or elsewhere, I should duplicate things to be able to compare a before-and-after result of this `reinit` command?

**Comment by seancpeters** [comment]
@patridge - All the configuration for dotnet new is located in `%USERPROFILE\.templateengine\dotnetcli\` for Windows users. There may be multiple subdirectories, each will be named for the version of dotnet they're for. So you should have a v2.0.0 folder - which contains all the relevant files.

**Comment by patridge** [comment]
@seancpeters After the `reinit` call, I was able to install and uninstall successfully. It did require a fully-qualified path. It also required I delete the terminating `\` from my full tab-completed path in PowerShell to complete successfully. So, there may be some value in communicating that in the docs and/or making the path input slightly more flexible.

As for the differences, somehow the old version I backed up from `%USERPROFILE%`, it **did not** contain the path I was trying to uninstall, and yet it was still present in the list of templates. I can't help but wonder if I had multiple versions in there somehow and it was masking the duplicate with a single entry.
